File: src/main/java/legacy/AppliedDiscountRepository.java
package legacy;

import java.util.List;
import java.util.Map;
import java.util.Optional;

public interface AppliedDiscountRepository {

    void add(AppliedDiscount discount);
    List<AppliedDiscount> findAll();
    List<AppliedDiscount> findForUser(UserId userId);
    Map<String, List<AppliedDiscount>> findForUserGroupedByType(UserId userId);
    Optional<AppliedDiscount> findLastAppliedDiscountForUser(UserId userId);
}

File: src/main/kotlin/refactored/AppliedDiscountRepository.kt
package refactored

interface AppliedDiscountRepository {
    fun add(discount: AppliedDiscount) : Unit
    fun findAll(): List<AppliedDiscount>
    fun findForUser(userId: UserId): List<AppliedDiscount>
    fun findForUserGroupedByType(userId: UserId): Map<String, List<AppliedDiscount>>
    fun findLastAppliedDiscountForUser(userId: UserId): AppliedDiscount?
}